src/cosl/coordinated_workers/worker.py

```diff
@@ -548,71 +547,68 @@ def _update_worker_config(self) -> bool:
+    def _sync_tls_files(self, tls_data: TLSData):
+        logger.debug("tls config in cluster. writing to container...")
+        if tls_data.privkey_secret_id:
+            private_key_secret = self.model.get_secret(id=tls_data.privkey_secret_id)
             private_key = private_key_secret.get_content().get("private-key")
-
-            ca_cert = tls_data["ca_cert"]
-            server_cert = tls_data["server_cert"]
-
-            # Read the current content of the files (if they exist)
-            current_server_cert = (
-                self._container.pull(CERT_FILE).read() if self._container.exists(CERT_FILE) else ""
-            )
-            current_private_key = (
-                self._container.pull(KEY_FILE).read() if self._container.exists(KEY_FILE) else ""
-            )
-            current_ca_cert = (
-                self._container.pull(CLIENT_CA_FILE).read()
-                if self._container.exists(CLIENT_CA_FILE)
-                else ""
-            )
-
-            if (
-                current_server_cert == server_cert
-                and current_private_key == private_key
-                and current_ca_cert == ca_cert
-            ):
-                # No update needed
-                return False
-
-            # Save the workload certificates
-            self._container.push(CERT_FILE, server_cert or "", make_dirs=True)
-            self._container.push(KEY_FILE, private_key or "", make_dirs=True)
-            self._container.push(CLIENT_CA_FILE, ca_cert or "", make_dirs=True)
-            self._container.push(ROOT_CA_CERT, ca_cert or "", make_dirs=True)
-
-            # Save the cacert in the charm container for charm traces
-            ROOT_CA_CERT.write_text(ca_cert)
         else:
+            private_key = None
+
+        new_contents: Optional[str]
+        any_changes = False
+        for new_contents, file in (
+            (tls_data.ca_cert, CLIENT_CA_FILE),
+            (tls_data.server_cert, CERT_FILE),
+            (private_key, KEY_FILE),
+            (tls_data.s3_tls_ca_chain, S3_TLS_CA_CHAIN_FILE),
+        ):
+            if not new_contents:
+                if self._container.exists(file):
+                    any_changes = True
+                    self._container.remove_path(file, recursive=True)
+                    logger.debug(f"{file} deleted")
+                    continue
+
+                logger.debug(f"{file} skipped")
+                continue
+
+            if self._container.exists(file):
+                current_contents = self._container.pull(file).read()
+                if current_contents == new_contents:
+                    logger.debug(f"{file} unchanged")
+                    continue
+
+            logger.debug(f"{file} updated")
+            any_changes = True
+            self._container.push(file, new_contents, make_dirs=True)
+
+        # Save the cacert in the charm container for charm traces
+        if tls_data.ca_cert:
```

Comment:
This could also be checked for changes before writing.

Reply:
I don't think it's worth the extra complexity, it's not a lot of IO